src/meter_recognition.py

```diff
@@ -1,14 +1,146 @@
+                sound.duration_signature = "1."
+            else:
+                sound.duration_signature = "64"
+        else:
+            sound.duration_signature = beat_frac_to_duration_sign[sound.beat_fraction]
+
+
+def beat_id_closest_to_timestamp(beats, timestamp):
+    clos = 1000
+    res = -1
+    for i in range(len(beats)):
+        if abs(beats[i]-timestamp) < clos:
+            clos = abs(beats[i]-timestamp)
+            res = i
+
+    return res
+
+
+def simple_sound_beat_dur(beats, sound):
+    id = sound.beat_id
+    b1 = beats[id]
+    b2 = beats[id+1]
+    sixteenth = (b2-b1)/4
+    return round(sound.duration/sixteenth)
+    # if abs(sound.timestamp-b1) < 0.05:
+    #     # super
+
+
+def tim_to_duration(tim):
+    result = []
+    for num in [32, 16, 8, 4, 2, 1]:
+        if tim >= num:
+            result.append(16//num)
+            tim -= num
+
+    # result1 = []
+    # for i in range(len(result)-1):
+    #     result1.append(str(result[i])+'~')
+    # result1.append(result[-1])
+    return result
+
+
+# def symbol_to_lilypond(symbol):
+#     s = symbol.lower().replace("#", "is")
+
+
+def update_sounds_brojaczj_algorithm(tempo, beats, sounds):
+    beats = list(beats)
+    while beats[0] > 0:
+        beats.insert(0, beats[0]-tempo)
+    while beats[-1] < sounds[-1].timestamp+sounds[-1].duration:
+        beats.append(beats[-1]+tempo)
+
+    dur = {}
+    for s in sounds:
+        id = f'{s.duration:.3f}'
+        dur[id] = dur.get(id, 0) + 1
+
+    for i in range(len(sounds)):
+        sounds[i].beat_id = beat_id_closest_to_timestamp(
+            beats, sounds[i].timestamp)
+            
+    for s in sounds:
+        duration = tim_to_duration(simple_sound_beat_dur(beats, s))
+        if (len(duration) >= 3) and \
+            (duration[1] == duration[0]*2) and \
+                (duration[2] == duration[1]*2):
+                    s.duration_signature=str(duration[0]//2)+'.'
+            
+        elif (len(duration) >= 2) and (duration[1] == duration[0]*2):
+            s.duration_signature=str(duration[0])+'.'
+        else:
+            s.duration_signature=str(duration[0])
+
+
+def update_sounds_with_beat_fractions_compare_adjacent(sounds, meter):
+    start_index = min(range(len(sounds)), key=[abs(s.duration-meter) for s in sounds].__getitem__)
+    if sounds[start_index].duration/meter > 1.5 or sounds[start_index].duration/meter < 0.75:
+        raise("Wow, there was no note with duration similar to the beat")
+    sounds[start_index].beat_fraction = 1.0
+    allowed_fractions = [16.0, 12.0, 8.0, 6.0, 4.0, 3.0, 2.0, 1.5, 1.0, 0.75, 0.5, 0.375, 0.25, 0.1875, 0.125, 0.09375, 0.0625]
+    allowed_fractions_log = [log2(ar) for ar in allowed_fractions]
+    for i in range(start_index+1, len(sounds)):
+        best_match_index = find_best_fraction_index(sounds[i], sounds[i-1], allowed_fractions_log)
+        sounds[i].beat_fraction = allowed_fractions[best_match_index]
```

Comment:
This is some relict of an old approach, beat_fraction should not be a Sound property - it is used only locally and introduce terrible ambiguity later on

Reply:
Fixed